[naga wgsl-in] Handle struct types in error messages properly.

jimblandy opened the pull request:
Don't try to use `&TypeInner` values to generate error messages, since `&TypeInner` cannot represent struct types in WGSL - they must be referred to by name, and only `Type` knows the type's name. Using `Handle<Type>` or `TypeResolution` works fine, so use that instead.

Add a new `Display` impl for `DiagnosticDisplay` for `TypeResolution`.

Fixes #7495.

jimblandy committed:
[naga wgsl-in] Handle struct types in error messages properly.

Don't try to use `&TypeInner` values to generate error messages, since
`&TypeInner` cannot represent struct types in WGSL - they must be
referred to by name, and only `Type` knows the type's name. Using
`Handle<Type>` or `TypeResolution` works fine, so use that instead.

Add a new `Display` impl for `DiagnosticDisplay` for `TypeResolution`.

Fixes #7495.

ErichDonGubler commented on naga/src/front/wgsl/lower/mod.rs:
```diff
@@ -3026,13 +3026,14 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
-                "Supplying argument {arg_index} of type {}",
-                crate::common::DiagnosticDisplay((ty, ctx.module.to_ctx()))
+                "Supplying argument {arg_index} of type {:?}",
+                crate::common::DiagnosticDebug((arg_type_resolution, &ctx.module.types))
```
thought: What, no `for_debug` here? 👀

jimblandy replied:
I changed this to use `ForDebugWithTypes`. Thanks.

ErichDonGubler commented on naga/src/common/diagnostic_display.rs:
```diff
@@ -49,6 +49,17 @@ use core::fmt;
+    }
+}
+
 impl fmt::Display for DiagnosticDisplay<(Handle<Type>, GlobalCtx<'_>)> {
```
question: `DiagnosticDisplay` with a `Handle<Type>` has a footgun in that it can't handle `struct`s. Should we try to remove this `Display` implementation, and force other use sites to migrate?

jimblandy replied:
It's `&TypeInner` that has the footgun, actually, not `Handle<Type>`.

Let me see if there are impls we can delete.

jimblandy replied:
This was a serious problem. I've greatly expanded the PR.

ErichDonGubler reviewed (approved):
Fixes a crash, so you've got my vote! Just one question that need not block.

jimblandy committed:
[naga wgsl-in] Don't panic printing struct types for conversions.

When a type conversion expression is ill-formed and the operand is a
struct type, don't panic trying to format the operand's type for the
error message.

jimblandy commented:
There are actually a lot of different places where this occurs, and I've been able to trigger crashes in most cases that I've found. I'm going to broaden the scope. I'll take it out of draft when I've taken care of all the cases I'm aware of.

jimblandy committed:
[naga wgsl-in] Don't panic printing struct types for initializers.

When an initializer has the wrong type and it's a struct, don't panic
trying to format its type for the error message.

jimblandy committed:
[naga wgsl-in] Avoid crashes when logging types.

In `naga::front::wgsl::lower`, change logging in
`ExpressionContext::automatic_conversion_consensus` to use
`TypeResolution`s instead of `TypeInner`s, so that structs don't cause
a crash.

jimblandy committed:
[naga] Remove `TypeContext::type_inner_to_string`.

In `naga::common::wgsl`, delete the method
`TypeContext::type_inner_to_string`, since it's a footgun: anyone
trying to convert a `TypeInner::Struct` to a string will hit
"unreachable" code.

jimblandy committed:
[naga wgsl-out] Stop using `TypeContext::write_type_inner`.

In `naga::back::wgsl`, prefer `TypeContext::write_type_resolution`
over `write_type_inner`, since the former is actually what we need,
and the latter can be troublesome if asked to write a `struct` type.

jimblandy committed:
[naga wgsl] New `TypeContext` method `write_unnamed_struct`.

When asked to generate WGSL for `TypeInner::Struct`, rather than
unconditionally calling `unreachable!`, defer to a new `TypeContext`
method, `write_unnamed_struct`.

Provide appropriate `write_unnamed_struct` implementations:

- In the WGSL backend, implement this as `unreachable!`, since the WGSL
  backend should always know the proper name to use for a struct.

- For diagnostic messages, generate something human-readable that
  indicates that some struct type was encountered.

- For logging and debugging, defer to `TypeInner`'s `Debug`
  implementation.

jimblandy committed:
[naga] Use `TypeContext::write_type_resolution` where appropriate.

Simplify the implementation of `Display` for
`DiagnosticDisplay<(TypeResolution, GlobalCtx)>` by calling
`GlobalCtx`'s implementation of `write_type_resolution` directly,
rather than duplicating its code. Provide a fallback for non-wgsl
builds.

jimblandy committed:
[naga] Delete `DiagnosticDisplay` impl for for `TypeInner`.

Delete the implementation of `core::fmt::Display` for
`naga::common::DiagnosticDisplay`, as it is a footgun wherever
`Struct` types can arise.

Document that it should not be implemented for `TypeInner`.

Although it is possible to put a dummy implementation in the way of
people adding real implementations, that would also turn attempts to
use it into dynamic errors rather than compile-time errors. Make do
with a comment for now.

jimblandy commented:
@ErichDonGubler Okay, the new version of this patch goes much further. I've deleted all the `TypeInner` formatting paths except the ones used for logging, and added tests for a few more cases that would panic.

Could you take another look?

jimblandy commented:
@ErichDonGubler Note that, aside from changing the original commit to use `for_debug` as you suggested, this leaves the original PR alone, and only adds commits. The new commits should be reviewable individually.

ErichDonGubler commented on naga/src/back/wgsl/writer.rs:
```diff
@@ -481,15 +481,19 @@ impl<W: Write> Writer<W> {
             module,
             names: &self.names,
         };
-        type_context.write_type_inner(inner, &mut self.out)?;
```
question(non-blocking): Would it be possible to retire `write_type_inner` entirely at some point?

jimblandy replied:
I started out with that as my goal. But one of the clienteles we need to serve is people doing debugging, and in some spots all you've got is a `TypeInner` and an arena.

Or at least I think so. Careful examination may show that, in every place we are currently dumping `TypeInner`, we actually have a `Handle<Type>` or a `TypeResolution` available as well. If that's the case, then we can totally make `write_type_inner` a private detail of `common::wgsl::types`.
